### ISSUE TITLE
Provide a 'feature("featureName", "g:a")' dependency declaration syntax in the spirit of 'testFixtures("g:a")'

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#25629]
### Expected Behavior

If I depend on the [test fixtures](https://docs.gradle.org/current/userguide/java_testing.html#consuming_test_fixtures_of_another_project) feature of a component, which is a [feature](https://docs.gradle.org/current/userguide/feature_variants.html) called `testFixtures`, I can do it like this:

```
dependencies {
   implementation(testFixtures(project(":module-a")))
}
``` 

If I depend on another feature variant that I defined and named myself – via `java.registereFeature("extraFun")` – I would like to be able to address that in a similar matter in a dependency declaration. For example:

```
dependencies {
   implementation(feature("extraFun", project(":module-a")))
}
```

### Context

Currently, you have to use the quite verbose `requireCapabilities` notation, for which you need to use/know things you have not defined yourself – the _group_ in the capability coordinates and how the "feature name" maps to the name in the coordinates:

```
dependencies {
   implementation(project(":module-a")) {
     capabilities {
       requireCapabilities("org.example:module-a-extra-fun")
     }
  }
}
```

But if we think in Java features, the underlying concept of capabilities should not need to shine through on the _consumer_ side. As it is also not required in the `registerFeature()` on the _producer_ side.

And it seems unintuitive that this is available for `testFixtures`, a "Gradle-owned feature", but not for custom features.

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

**Comment by jvandort** [comment]
Since this issue does not exactly match the feature added (but the new feature solves the use case) we have written a separate issue for the release notes entry https://github.com/gradle/gradle/issues/30685

**Comment by jjohannes** [comment]
I am very happy to see a solution for this implemented. Thank you @jvandort for adding the feature!

**Comment by jjohannes** [comment]
I just got around to trying this out now. I am happy that there is such a feature now, but why was the original syntax suggestion rejected?

It's now still much too noisy IMO, because you need to add all these curly braces. And you still see the word `capabilities` (which you don't see on the producer side). 😒 

```
    implementation(testFixtures(projects.app))
    implementation(projects.app) { capabilities { requireFeature("mock-api") } }
```

I was hoping for something like this:

```
    implementation(testFixtures(projects.app))
    implementation(feature("mock-api", projects.app))
```

**Comment by leonard84** [comment]
I agree with @jjohannes. Having to use capabilities feels clunky.

**Comment by jvandort** [comment]
I agree the implemented API vs the API proposed in this issue is clunkier. 

Though I question, how would you implement `feature("mock-api", projects.app)`? 

We call methods like this "dependency modifiers". Some common ones are:

1. The `platform` / `enforcedPlatform` modifiers:
`platform(projects.app)` - This mutates the passed-in dependency and adds an attribute to it. Under the hood, this is a light wrapper around:

```
implementation(projects.app) {
    attributes {
        attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.REGULAR_PLATFORM))
    }
    endorseStrictVersions()
}
```

2. The `test-fixtures` modifier
`testFixtures(projects.app)` - This mutates the passed-in dependency and adds a capability to it. Before this feature was implemented, it was a light wrapper around:

```
implementation(projects.app) {
    capabilities {
        requireCapability(new ProjectDerivedCapability(<above_dependency>.dependencyProject, "testFixtures"))
    }
}
```

In this case, we use the internal `ProjectDerivedCapability` to create a special capability that at resolution-time resolves to the coordinates of the target project (in an IP unsafe manner).

The core point here is that (without wrapping the dependency in some other wrapper type that later needs to be `instanceof`'d and deconstructed), we need to have some lower-level construct on the dependency that the modifier can call into.

Without the feature as implemented, I don't see how we would implement the feature as requested in this issue.

**Comment by jjohannes** [comment]
I am not sure I fully grasp the implementation details. My thought was, from a naive usability perspective, if there is a

```
implementation(testFixtures(projects.app))
```

It should also be possible to have a

```
implementation(feature("mock-api", projects.app))
```

-----

I also didn't mean to say that the new technical solution is bad. I think it is great! I am just saying that the user facing API could be improved. Can't you add the original proposed API now on top of the new implementation? I see that `testFixtures` has been modified to work like this now:

```
ModuleDependency moduleDependency = (ModuleDependency) testFixturesDependency;
moduleDependency.capabilities(c -> c.requireFeature(TEST_FIXTURES_CAPABILITY_FEATURE_NAME));
```

I think an API like `feature("...", projects.app)` could be implemented similarly (or am I missing something?).

```
ModuleDependency moduleDependency = (ModuleDependency) featureDependency;
moduleDependency.capabilities(c -> c.requireFeature(featureName)); // 'featureName' being the 1st method parameter
```

**Comment by hfhbd** [comment]
I also prefer the notation of @jjohannes so I created a PR to add this overload: https://github.com/gradle/gradle/pull/31358

**Comment by jvandort** [comment]
Reopened since the original fix did not add a convenient API as proposed. Lets fix this for with https://github.com/gradle/gradle/pull/31358

**Comment by xenoterracide** [comment]
is `feature(...` meant to be the actual name of the method?

```kts
val demoServer by sourceSets.creating

java {
  registerFeature("demoServer") {
    usingSourceSet(demoServer)
  }
}

val demoServerImplementation by configurations.existing
val demoServerRuntimeOnly by configurations.existing
val demoServerApi by configurations.existing
```

`demoServer` here is the name of my feature and source set... but probably also what I'd like the name of my method to be

what'd really be nice is a more intuitive less verbose DSL

```kts
val demoServer by java.registerFeature // automatically uses source set of the same name
val demoServerImplementation by configurations.existing
val demoServerRuntimeOnly by configurations.existing
val demoServerApi by configurations.existing

dependencies { implementation(demoServer(project)) }
```

**Comment by hfhbd** [comment]
@xenoterracide This issue is about the consumer project of another project with a registered feature:

```kotlin
// lib.gradle.kts
java.registerFeature("myFeature") {
    usingSourceSet(sourceSets.create("myFeature"))
}

// lib2.gradle.kts
dependencies {
    implementation(project(":lib")) {
        capabilities {
            requireFeature("my-feature")
        }
    }
}
```

But I also agree, for easier declaring dependencies, the FeatureSpec should provide a custom jvm dependency handler:
```kotlin
// lib.gradle.kts
java.registerFeature("myFeature") {
    usingSourceSet(sourceSets.create("myFeature")) // and maybe always register automatically sourceSet if no one is set explicitly
    dependencies {
        implementation("com.example:foo:1.2.3")
        api("com.example:foo:1.2.3")
        // etc...
    }
}
```
But this should be a separate issue.

**Comment by xenoterracide** [comment]
The suggestion here. Is that 

implementation(feature("mock-api", projects.app))

I am saying that should be 

implementation(mockApi(projects.app))

The whole test fixtures thing works in exactly the same way as what I showed. Except we aren't getting the fancy method.

**Comment by hfhbd** [comment]
@xenoterracide How exactly do you plan to fetch and generate a type safe accessor in the consumer project, here lib2? The feature is registered in the producer project, here lib. lib could be also published.

`testFixtures` is different because it is hardcoded.

**Comment by xenoterracide** [comment]
That's a good question. When I wrote my response comment that problem hadn't quite clicked to me. However, I think that should be part of the problem that's trying to be solved here. Perhaps even if it's only just kotlin DSL... Because the stringy thing makes me unhappy especially needing to be repeatedly reused. 

It might be reasonable to ask how does the projects.app accessor work? That's not hand coded... Perhaps there's room to expand that to 

implementation(projects.app.mockApi)

I'm spitballing but I don't feel like the idea that it's a separate sub project is a factoring limitation here
